unyt/dask_array.py

```diff
@@ -0,0 +1,620 @@
+
+
+def _create_with_quantity(dask_array, new_unyt_array):
+    """
+    this function instantiates a new unyt_dask_array instance and then sets
+    the unit state, including units. Used to wrap dask operations
+
+    Parameters
+    ----------
+    dask_array : a standard dask array
+    new_unyt_array : a standard unity array
+    remaining arguments get passed to unyt.unyt_array, check there for a
+    description.
+    """
+    out = unyt_from_dask(dask_array)
+
+    # attach the unyt_array
+    units = new_unyt_array.units
+    unyt_name = new_unyt_array.name
+
+    out._set_unit_state(units, new_unyt_array, unyt_name)
+    return out
+
+
+def unyt_from_dask(
+    dask_array,
+    units=None,
+    registry=None,
+    bypass_validation=False,
+    unyt_name=None,
+):
+    """
+    creates a unyt_dask_array from a standard dask array.
+
+    Parameters
+    ----------
+    dask_array : a standard dask array
+
+    remaining arguments get passed to unyt.unyt_array, check there for a
+    description.
+
+    Examples
+    --------
+
+    >>> from unyt import dask_array, unyt_quantity
+    >>> import dask.array as da
+    >>> x = da.ones((1000, 1000), chunks=(100, 100))
+    >>> x_da = dask_array.unyt_from_dask(x, 'm')
+    >>> type(x_da)
+    <class 'unyt.dask_array.unyt_dask_array'>
+    >>> x_da.units
+    m
+    >>> x_da.mean().units
+    m
+    >>> x_da.mean().compute()
+    unyt_quantity(1., 'm')
+    >>> x_da.to('cm').mean().compute()
+    unyt_quantity(100., 'cm')
+    >>> (x_da.to('cm')/unyt_quantity(10, 's')).mean().compute()
+    unyt_quantity(10., 'cm/s')
+
+    """
+
+    # reduce the dask array to pull out the arguments required for instantiating
+    # a new dask.array.core.Array object and then initialize our new unyt_dask
+    # array
+    (cls, args) = dask_array.__reduce__()
+
+    da = unyt_dask_array(
+        *args,
+        units=units,
+        registry=registry,
+        bypass_validation=bypass_validation,
+        unyt_name=unyt_name,
+    )
+
+    return da
+
+
+# note: the unyt_dask_array class has no way of catching daskified reductions.
+# operations like dask.array.min() get routed through dask.array.reductions.min()
+# and will return plain arrays or float/int values.
+#
+# When these operations exist as attributes, they can be called and will return
+# unyt objects. i.e., :
+#
+# import dask; import unyt
+# x_da = unyt_from_dask(dask.array.ones((10, 10), chunks=(2, 2)), unyt.m)
+# dask.array.min(x_da).compute()  #  returns a plain float
+# x_da.min().compute()  #  returns a unyt quantity
+#
+# but when the functions do not exist as attributes, like dask.array.nanmin(),
+# it is difficult to handle without manually wrapping all of those reductions
+# functions and exposing them here. The following function, reduce_with_units,
+# is a compromise: it is a simple helper function for calling
+# a general dask.array.reductions method on a unyt_dask array to correctly
+# handle units.
```

Comment:
very useful comment, I would suggest to convert it to an issue and link it here, so there's a better chance of finding it back in the future (IMO)